src/awkward/_v2/_connect/rdataframe/from_rdataframe.py

```diff
@@ -94,61 +94,70 @@ def form_dtype(form):
             else:
                 return form_dtype(form.content)
 
-        dtype = form_dtype(form)
-        buffers = {}
-        depths = 0
-        offsets_length = 0
+        def empty_buffers(cpp_buffers_self, names_nbytes):
+            buffers = {}
+            for item in names_nbytes:
+                buffers[item.first] = ak.nplike.numpy.empty(item.second)
+                cpp_buffers_self.append(
+                    item.first,
+                    buffers[item.first].ctypes.data_as(ctypes.POINTER(ctypes.c_ubyte)),
+                )
+            return buffers
+
+        data_type = cpp_type_of[form_dtype(form).name]
 
         # pull in the CppBuffers (after which we can import from it)
-        CppBuffers = cppyy.gbl.awkward.CppBuffers[column_type, cpp_type_of[dtype.name]]
+        CppBuffers = cppyy.gbl.awkward.CppBuffers[column_type]
         cpp_buffers_self = CppBuffers(result_ptrs)
 
         if isinstance(form, ak._v2.forms.NumpyForm):
-            distance = CppBuffers.result_distance(cpp_buffers_self)
-            data = ak.nplike.numpy.empty(distance, dtype)
-            CppBuffers.fill_data_array(
-                cpp_buffers_self, data.ctypes.data_as(ctypes.c_void_p)
-            )
-            layout = ak._v2.contents.numpyarray.NumpyArray(
-                data,
-                parameters=form.parameters,
-            )
-
-            return _wrap_as_record_array(layout)
+
+            NumpyBuilder = cppyy.gbl.awkward.LayoutBuilder.Numpy[data_type]
+            builder = NumpyBuilder()
+            builder_type = type(builder).__cpp_name__
+
+            cpp_buffers_self.fill_from[builder_type](builder)
 
         elif isinstance(form, ak._v2.forms.ListOffsetForm) and isinstance(
             form.content, ak._v2.forms.NumpyForm
         ):
-            # list_depth == 2 or 1 if its the list of strings
-            # copy data from RDF and make nested offsets
-            depths, offsets_length = CppBuffers.offsets_and_flatten_2(cpp_buffers_self)
+            # NOTE: list_depth == 2 or 1 if its the list of strings
+            ListOffsetBuilder = cppyy.gbl.awkward.LayoutBuilder.ListOffset[
+                "int64_t",
+                f"awkward::LayoutBuilder::Numpy<{data_type}",
+            ]
+            builder = ListOffsetBuilder()
+            builder_type = type(builder).__cpp_name__
+
+            cpp_buffers_self.fill_offsets_and_flatten_2[builder_type](builder)
 
         elif list_depth == 3:
-            depths, offsets_length = CppBuffers.offsets_and_flatten_3(cpp_buffers_self)
+            ListOffsetBuilder = cppyy.gbl.awkward.LayoutBuilder.ListOffset[
+                "int64_t",
+                f"awkward::LayoutBuilder::ListOffset<int64_t, awkward::LayoutBuilder::Numpy<{data_type}>",
+            ]
```

Comment:
I think, this should be generated from a Form. I'll do it as a separate PR after introducing multiple columns.